Api Task

ofekcMoveo opened the pull request:
add Request struct that represents a generic http request
add structs that contains the JSON objects as received from the api 
change to MVVM-
1. add HomeControllerViewModel which holds the articles array and getting data from models,
2. add Repositories folder with ArticleRepository.swift with manages the building and the sending of the request with Alamofire,
3. add Networking folder with AlamofireManager.swift which only sends a request
4. add methods to HomepageController in order to support the MVVM model and get the data from the modelView

add loader to screen while fetching data from api

ofekcMoveo committed:
add Request struct that represents a generic http request
add structs that contains the JSON objects as recived from the api
change to MVVM-
1. add HomeControllerViewModel,
2. add Repositories folder with ArticleRepository.swift with manages the building and the sending of the request with alamofire,
3. add Networking folder with AlamofireManager.swift which sends a request
4. add methods to HomepageController in order to support the MVVM model and get the data from the modelView

add loader to screen while fetching data from api

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+struct AppConstants {
```
I would change the file and the struct to "CellConstants" and have only cell-related stuff in here

noamkuMoDev replied:
Also I would suggest having two separate structs inside this file - struct TableCellsIdentifier & struct NibNames

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -23,11 +26,42 @@ class HomeViewController: UIViewController {
         articlesTableView.delegate = self
         
-        articlesTableView.register(UINib(nibName: "ArticleCell", bundle: nil), forCellReuseIdentifier: ArticleCell.articleCellIdentifier)
+        articlesTableView.register(UINib(nibName: "ArticleCell", bundle: nil), forCellReuseIdentifier: AppConstants.articleCellIdentifier)
```
We'd want to save the "ArticleCell" string as a constant and not write it manually when we need it

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -23,11 +26,42 @@ class HomeViewController: UIViewController {
+        self.activityIndicator.startAnimating()
+        homepageViewModel.getArticlesFromAPI(completionHandler: { errorMsg in
+            if(errorMsg != nil) {
+                self.showErrorAlert((errorMsg!))
```
Forgot to stop the activityIndicator in case of an error

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -94,6 +127,13 @@ extension HomeViewController: UITableViewDataSource {
         return 40
     }
     
+    private func formatDate(_ date: String) -> String {
```
It's better practice to create this function inside a util file because we might need this function elsewhere and we don't want to write the same code more than once

noamkuMoDev commented on Dispatcher/Networking/AlamofireManager.swift:
```diff
@@ -0,0 +1,25 @@
+
+class AlamofireManager {
+    
+    func SendRequest<T:Decodable>(_ request: Request, completionHandler: @escaping (Result<T, Error>) -> Void) {
```
function name should start with a small letter

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,41 @@
+            case .success(let dataResult):
+                for article in dataResult.articles {
+                    if(article.language == "en") {
+                        let article = Article(id: article.id, title: article.title, summary: article.summary, author: article.author, topic: [article.topic], imageURL: article.media, language: article.language, date: article.publishedDate)
```
for better readability personally I prefer to hit enter and have the Article object in multiple lines

noamkuMoDev commented on Dispatcher/Models/UserInput.swift:
```diff
@@ -0,0 +1,34 @@
+
+import Foundation
+
+struct UserInput: Decodable {
```
What is this model for?

ofekcMoveo committed:
fix readability on the code,
add utils file,
stop animating activityIndicator if there is error

noamkuMoDev commented on Dispatcher/Supporting Files/Utils.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+struct Utils {
```
we don't really have to create a struct in order to have util functions. We can just write "func formatDate" directly in the swift file and spare the struct

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,49 @@
+    var articles: [Article] = []
+    
+    func getData(completionHandler: @escaping (_ articles: [Article], _ errorMsg: String?) -> Void) {
+        let request = Request(baseUrl: APIConstants.newscatcherURL, headers: ["x-api-key": APIConstants.APIkey], parameters: ["countries" : "IL"], method: .get)
```
Since we'll have the same headers, parameters and method - we wouldn't want to write the same initialization of Request over and over again for every request that we will want to send out.

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,49 @@
+    let alamofireManager = AlamofireManager()
+    var articles: [Article] = []
+    
+    func getData(completionHandler: @escaping (_ articles: [Article], _ errorMsg: String?) -> Void) {
```
"getData" is not a good function name as it is not specific enough

noamkuMoDev commented on Dispatcher/Repositories/ArticlesRepository.swift:
```diff
@@ -0,0 +1,49 @@
+class ArticlesRepository {
+
+    let alamofireManager = AlamofireManager()
+    var articles: [Article] = []
```
No need to save the articles in the repository as we pass them back to the viewModel

noamkuMoDev commented on Dispatcher/Constants/APIConstants.swift:
```diff
@@ -0,0 +1,13 @@
+
+struct APIConstants {
+    static let newscatcherURL = "https://api.newscatcherapi.com/v2/latest_headlines"
+    static let APIkey = "sOEtokE8TQCbCtBWifqf87OgISACtNadOaVf88T9vAY"
```
It is not safe to store sensitive data such as API keys like this, especially as it is being uploaded to github and is available for everyone to abuse.
A better solution - create a JSON file (e.g. Config.json) to store the key inside. To use the key in your project just read the API key value from the file. Don't forget to add the json file into your .gitignore so it won't be uploaded to github.

noamkuMoDev commented on Dispatcher/Constants/AppConstants.swift:
```diff
@@ -0,0 +1,13 @@
+
+struct AppConstants {
+    static let articleCellIdentifier = "articleCell"
+    static articleCellNibName = "ArticleCell"
```
forgot the "let" in "static let articleCellNibName".

noamkuMoDev commented:
Let's place the Assets, Info, CoreData, AppDelegate and SceneDelegate inside the "Supporting Files" folder instead of having them floating around in our project's hierarchy. Will be more organized this way

noamkuMoDev commented:
Inside the Assets - let's place all colors inside a "Colors" folder and the tab-nav icons inside a dedicated folder as well so we are more organized

noamkuMoDev commented on Dispatcher/Views/TableViewCells/ArticleCell.xib:
```diff
@@ -13,7 +13,7 @@
         <tableViewCell clipsSubviews="YES" contentMode="center" selectionStyle="none" indentationWidth="10" reuseIdentifier="articleCell" rowHeight="460" id="KGk-i7-Jjw" customClass="ArticleCell" customModule="Dispatcher" customModuleProvider="target">
```
You have a little issue with the constraints in this file: 
![Screen Shot 2023-05-05 at 14 50 23](https://user-images.githubusercontent.com/100192030/236450391-76470586-ddf0-4d83-9c08-87ab91122001.png)

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -23,11 +26,43 @@ class HomeViewController: UIViewController {
         articlesTableView.dataSource = self
```
We can place all of this tableview-related-stuff inside a dedicated function like "setupTableView"

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -23,11 +26,43 @@ class HomeViewController: UIViewController {
+        homepageView.addSubview(activityIndicator)
+    }
+    
+    private func fetchArticles() {
```
Following the CleanCode principles, we should stick to a single phrase for describing the action of retrieving data from the api. Either choose "fetch" or "get", but make sure to use the chosen word throughout the project in all related function names

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -93,7 +127,6 @@ extension HomeViewController: UITableViewDataSource {
     func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
         return 40
```
Better save this value as a variable in a dedicated constants file so it's easier to change / replicate on other pages

noamkuMoDev commented:
Inside the HomeViewController.swift -> inside viewForHeaderInSection func -> you have:
       view.backgroundColor = UIColor(named: "screenBackgroundColor")
It's better to have the colorName as a a variable in a dedicated constants file so that if the name needs to be changed in the future - we need to change only in one place.

Also, under it you have:
       label.text = "Top Headlines in Israel"
Add a dedicated file for the hard-coded texts

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+        return articlesToDisplay.count
+    }
+    
+    func currentArticle(index: Int) -> Article {
```
It's not exactly returning the currentArticle necessarily, so I think a better name would be "getArticleByIndex(_ index: Int)" as this is what the function actually does

noamkuMoDev commented on Dispatcher/Controllers/Homepage/HomepageViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+//
+
+import Foundation
+import UIKit
```
In any file that we don't actually use UI elements - we can spare importing the UIKit.
Importing unnecessary modules can increase the build time and compile time, as the Swift compiler needs to load and process more code

abenartz commented on Dispatcher/Models/ArticleResponse.swift:
```diff
@@ -0,0 +1,52 @@
+    let authors: String
+    let media: String
+    let isOpinion: Bool
+    let twitterAccount: String?
```
Do you use all of the properties from the server in your app? Clean your objects by removing unnecessary values.

abenartz commented on Dispatcher/Models/ArticleResponse.swift:
```diff
@@ -0,0 +1,52 @@
+    let isOpinion: Bool
+    let twitterAccount: String?
+    let score: Double?
+    let id: String
```
It's common to put the id property first. Important to notice the object has an identifier.

abenartz commented on Dispatcher/Controllers/Homepage/HomeViewController.swift:
```diff
@@ -23,11 +26,43 @@ class HomeViewController: UIViewController {
+        homepageViewModel.getArticlesFromAPI(completionHandler: { errorMsg in
+            if(errorMsg != nil) {
+                self.showErrorAlert((errorMsg!))
+                self.activityIndicator.stopAnimating()
```
stopAnimating() happens both times. Can get out of the condition...

ofekcMoveo committed:
delete the Utils struct
fix functions names
clean the ArticleResponse object
create const var for tableRowHeight
create "setupTableView" func in homepage controller
fix constraints in ArticleCell
organize files in groups
